### DISCUSSION TITLE
Trouble Restricting Number of Tracks

### DISCUSSION THREAD

**Discussion by majaneubauer** [talmolab/sleap#1679] [Help!] [answered]
I am currently using Google Colab and SLEAP (Version 1.3.3) to train a bottom-up multi-animal model.  I have videos of two mice (one black and one white) in an open field arena, filmed from below. I am training my model so that it is able to identify the two animals in my video. I managed to export a training package to Google Colab where I specified the maximum number of instances to be 2. Below I have attached screenshots of my training pipeline and bottom-up model configuration:
![image](https://github.com/talmolab/sleap/assets/131380612/dc1dbccd-5bf2-43ce-957b-4713d5cd5867)
![image](https://github.com/talmolab/sleap/assets/131380612/43df383a-1f76-45ed-a02c-29265e57d55e)

This is what I did to train my model in Google Colab:
`!sleap-train multi_instance.json labels.v001_bottom_up.pkg.slp --run_name "240205_multi_instance"`

I am now trying to predict on a video that I have uploaded to my Google Drive. Since I have two animals I want to restrict my maximum instances to be 2 and also want to restrict my maximum number of tracks to be 2. Initially, I used 
`!sleap-track videos_to_analyse/Trial2.mp4 \
    --tracking.tracker simple \
    -m models/240205_multi_instance.multi_instance` 
to predict on my video. This left me with way too many tracks and also way too many instances. 

After having a look at the [Command Line Interface](https://sleap.ai/develop/guides/cli.html) section I found various arguments that I could use to restrict my model. I then tried various combinations of the arguments that I thought would restrict my model, for example the one below:

`!sleap-track videos_to_analyse/Trial2.mp4 \
    --max_instances 2 \
    --tracking.max_tracks 2 \
    --tracking.tracker flow \
    --tracking.target_instance_count 2 \
    --tracking.pre_cull_to_target 2 \
    --tracking.post_connect_single_breaks 2 \
    -m models/240205_multi_instance.multi_instance`

Additionally I tried to set `--tracking.max_tracking` to true. This left me with this error message:
Exception in thread Thread-1 (_object_builder):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 3245, in 
_object_builder
    predicted_instances = self.tracker.track(
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/tracking.py", line 688, in track
    candidate_instances = self.candidate_maker.get_candidates(
TypeError: FlowCandidateMaker.get_candidates() got an unexpected keyword argument 
'track_matching_queue_dict'

However, all the combinations I've tried made me end up with 8 tracks:
`!sleap-inspect videos_to_analyse/Trial2.mp4.predictions.slp`
INFO:numexpr.utils:NumExpr defaulting to 8 threads.
Labeled frames: 10956
Tracks: 8
Video files:
  videos_to_analyse/Trial2.mp4
    labeled frames: 10956
    labeled frames from 0 to 10955
    user labeled frames: 0
    tracks: 8
    max instances in frame: 2
Total user labeled frames: 0

Provenance:
  model_paths: ['models/240205_multi_instance.multi_instance/training_config.json']
  predictor: BottomUpPredictor
  sleap_version: 1.3.3
  platform: Linux-6.1.58+-x86_64-with-glibc2.35
  command: /usr/local/bin/sleap-track videos_to_analyse/Trial2.mp4 --max_instances 2 --tracking.tracker flow --tracking.target_instance_count 2 --tracking.pre_cull_to_target 2 --tracking.post_connect_single_breaks 1 -m models/240205_multi_instance.multi_instance
  data_path: videos_to_analyse/Trial2.mp4
  output_path: videos_to_analyse/Trial2.mp4.predictions.slp
  total_elapsed: 689.5119385719299
  start_timestamp: 2024-02-05 14:25:22.100585
  finish_timestamp: 2024-02-05 14:36:51.612507
  args: {'data_path': 'videos_to_analyse/Trial2.mp4', 'models': ['models/240205_multi_instance.multi_instance'], 'frames': '', 'only_labeled_frames': False, 'only_suggested_frames': False, 'output': None, 'no_empty_frames': False, 'verbosity': 'rich', 'video.dataset': None, 'video.input_format': 'channels_last', 'video.index': '', 'cpu': False, 'first_gpu': False, 'last_gpu': False, 'gpu': 'auto', 'max_edge_length_ratio': 0.25, 'dist_penalty_weight': 1.0, 'batch_size': 4, 'open_in_gui': False, 'peak_threshold': 0.2, 'max_instances': 2, 'tracking.tracker': 'flow', 'tracking.max_tracking': None, 'tracking.max_tracks': None, 'tracking.target_instance_count': 2, 'tracking.pre_cull_to_target': 2, 'tracking.pre_cull_iou_threshold': None, 'tracking.post_connect_single_breaks': 1, 'tracking.clean_instance_count': None, 'tracking.clean_iou_threshold': None, 'tracking.similarity': None, 'tracking.match': None, 'tracking.robust': None, 'tracking.track_window': None, 'tracking.min_new_track_points': None, 'tracking.min_match_points': None, 'tracking.img_scale': None, 'tracking.of_window_size': None, 'tracking.of_max_levels': None, 'tracking.save_shifted_instances': None, 'tracking.kf_node_indices': None, 'tracking.kf_init_frame_count': None}

This is what it looks like when I merge my predictions file into my initially labelled video:
![image](https://github.com/talmolab/sleap/assets/131380612/fed8b4be-b4ff-44bf-b16f-50c7f765f211)
This is already loads better than what I had before I used any restricting arguments during inference. From here on, I would be able to merge my tracks so that I end up with two animals and two tracks total. 

Ideally, I don't want to do this manually and rather have my tracks be restricted to 2 in the first place. Do you have any suggestions what I need to pass to which argument in `sleap-track` for this to work?

Thank you!

**Comment by roomrys** [comment, marked as answer]
As an update, the issue was resolved by @majaneubauer in #1676:
> I have been doing some testing on my own the past two days and just now realised where I went wrong. Exporting the training package specifying the max instances to 2 and predicting on user-labeled frames was not the problem. Rather, I made a mistake while running inference on Colab using `sleap-track`. I reported my issue in [#1679](https://github.com/talmolab/sleap/discussions/1679), which I have now closed. I had to specify my `tracking.tracker` to `flowmaxtracks` in order for `tracking.max_tracking` to work.

For all: `tracking.tracker` should be set to `flowmaxtracks` or `simplemaxtracks` to use the `tracking.max_tracks` argument.

I found a "`Hacky`" line in the GUI where we force `maxtracks` to be appended to whichever tracker was selected in the GUI: evidence that this was indeed the missing piece running `sleap-track` through the CLI.
https://github.com/talmolab/sleap/blob/5fa10795f51fd25568c444dd95d370b86054acdc/sleap/gui/learning/runners.py#L237-L245

Thanks,
Liezl
